File: ecommerce_api/src/main/java/personal/rajit/controller/PurchaseController.java
package personal.rajit.controller;

import lombok.AllArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import personal.rajit.controller.request_dto.PurchaseSearchDto;
import personal.rajit.entity.Purchase;
import personal.rajit.service.PurchaseService;

@RestController
@RequestMapping("/purchase")
@AllArgsConstructor
public class PurchaseController {
    private final PurchaseService purchaseService;

    @PostMapping("/save")
    public ResponseEntity<Purchase> save(@RequestBody Purchase purchase) {
        return new ResponseEntity(purchaseService.savePurchase(purchase), HttpStatusCode.valueOf(200));
    }

    @PostMapping("/update")
    public ResponseEntity<Purchase> update(@RequestBody Purchase purchase) {
        return new ResponseEntity(purchaseService.updatePurchase(purchase), HttpStatusCode.valueOf(200));
    }

    @PostMapping("/delete")
    public ResponseEntity<String> deletePurchase(@RequestBody Purchase purchase) {
        return new ResponseEntity(purchaseService.deletePurchase(purchase), HttpStatusCode.valueOf(200));
    }

    @PostMapping("/search")
    public ResponseEntity<Page<Purchase>> search(@RequestBody PurchaseSearchDto purchaseSearchDto) {
        return new ResponseEntity(purchaseService.searchPurchase(purchaseSearchDto), HttpStatusCode.valueOf(200));
    }

//    @PostMapping("/save-with-purchase-detail")
//    public ResponseEntity<Page<Purchase>> saveWithDetail(@RequestBody Purchase purchase) {
//        return new ResponseEntity(purchaseService.saveWithDetail(purchase), HttpStatusCode.valueOf(200));
//    }

}

File: frontend/src/app/controller/PurchaseController.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Purchase } from "../entity/Purchase";
import { PurchaseSearchDto } from "./search_dto/PurchaseSearchDto";
import { Page } from "../dto/Page";
import { PurchaseDetail } from "../entity/PurchaseDetail";

@Injectable({ providedIn: 'root' })
export class PurchaseController {

  constructor(private httpClient: HttpClient,) { }

  save(purchase: Purchase): Observable<Purchase> {
    return this.httpClient.post<Purchase>('http://localhost:8080/purchase/save',
      purchase);
  }

  update(purchase: Purchase): Observable<Purchase> {
    return this.httpClient.post<Purchase>('http://localhost:8080/purchase/update',
      purchase);
  }

  delete(purchase: Purchase): Observable<boolean> {
    return this.httpClient.post<boolean>('http://localhost:8080/purchase/delete',
      purchase)
  }

  search(purchaseSearchDto: PurchaseSearchDto): Observable<Page<Purchase>> {
    return this.httpClient.post<Page<Purchase>>('http://localhost:8080/purchase/search',
      purchaseSearchDto);
  }

  saveWithDetail(purchase: Purchase): Observable<Purchase> {
    return this.httpClient.post<Purchase>('http://localhost:8080/purchase/save-with-purchase-detail', purchase);
  }

  // searchWithDetailAndProduct(purchase: Purchase): Observable<Page<PurchaseDetail>>{
  //   return this.httpClient.post<PurchaseDetail>('http://localhost:8080/purchase/search-with-purchase-detail', purchase);
  // }

}